pageserver/src/tenant/timeline/detach_ancestor.rs

```diff
@@ -0,0 +1,159 @@
+    let mut rest_of_historic = vec![];
+
+    let mut later_by_lsn = 0;
+
+    for desc in source_layermap.layer_map().iter_historic_layers() {
+        // off by one chances here:
+        // - start is inclusive
+        // - end is exclusive
+        if desc.lsn_range.start > ancestor_lsn {
+            later_by_lsn += 1;
+            continue;
+        }
+
+        let target = if desc.lsn_range.start <= ancestor_lsn
+            && desc.lsn_range.end > ancestor_lsn
+            && desc.is_delta
+        {
+            // TODO: image layer at Lsn optimization
+            &mut straddling_branchpoint
+        } else {
+            &mut rest_of_historic
+        };
+
+        target.push(source_layermap.get_from_desc(&desc));
+    }
+
+    (later_by_lsn, straddling_branchpoint, rest_of_historic)
+}
+
+pub(super) async fn upload_rewritten_layer(
+    end_lsn: Lsn,
+    layer: &Layer,
+    target: &Arc<Timeline>,
+    cancel: &CancellationToken,
+    ctx: &RequestContext,
+) -> Result<Option<Layer>, DetachFromAncestorError> {
+    use DetachFromAncestorError::UploadRewritten;
+    let copied = copy_lsn_prefix(end_lsn, layer, target, ctx).await?;
+
+    let Some(copied) = copied else {
+        return Ok(None);
+    };
+
+    // FIXME: better shuttingdown error
+    target
+        .remote_client
+        .as_ref()
+        .unwrap()
+        .upload_layer_file(&copied, cancel)
+        .await
+        .map_err(UploadRewritten)?;
+
+    Ok(Some(copied.into()))
+}
+
+async fn copy_lsn_prefix(
+    end_lsn: Lsn,
+    layer: &Layer,
+    target_timeline: &Arc<Timeline>,
+    ctx: &RequestContext,
+) -> Result<Option<ResidentLayer>, DetachFromAncestorError> {
+    use DetachFromAncestorError::{CopyDeltaPrefix, RewrittenDeltaDownloadFailed};
+
+    tracing::debug!(%layer, %end_lsn, "copying lsn prefix");
+
+    let mut writer = DeltaLayerWriter::new(
+        target_timeline.conf,
+        target_timeline.timeline_id,
+        target_timeline.tenant_shard_id,
+        layer.layer_desc().key_range.start,
+        layer.layer_desc().lsn_range.start..end_lsn,
+    )
+    .await
+    .map_err(CopyDeltaPrefix)?;
+
+    // likely shutdown
+    let resident = layer
+        .download_and_keep_resident()
+        .await
+        .map_err(RewrittenDeltaDownloadFailed)?;
+
+    let records = resident
+        .copy_delta_prefix(&mut writer, end_lsn, ctx)
+        .await
+        .map_err(CopyDeltaPrefix)?;
+
+    drop(resident);
+
+    tracing::debug!(%layer, records, "copied records");
+
+    if records == 0 {
+        drop(writer);
+        // TODO: we might want to store an empty marker in remote storage for this
+        // layer so that we will not needlessly walk `layer` on repeated attempts.
+        Ok(None)
+    } else {
+        // reuse the key instead of adding more holes between layers by using the real
+        // highest key in the layer.
+        let reused_highest_key = layer.layer_desc().key_range.end;
```

Comment:
Comment is more distracting than helpful, it seems obvious to use the highest key here.
Why was it noteworthy in your opinion?

```suggestion
    } else {
        let reused_highest_key = layer.layer_desc().key_range.end;
```

Reply:
It is not necessarily the highest key we wrote to the layer.

Reply:
> we wrote to the layer.

to the NEW layer?

Reply:
I am unsure what are you pointing at with this:

> > we wrote to the layer.
> 
> to the NEW layer?

We currently do not know what was the highest key that we wrote to the layer. Consider a situation where we cut at lsn X, and the delta has keys:

- 0x0, X-2
- 0x1, X-1
- 0x1f, X-0
- 0x2, X+1
- 0x3, X+2

The old delta would have had a key range of 0..0x3, but the lens prefix copied has 0..0x1.

Do you think we should instead create non-partitioning based holes in between layers? The current method always preserves L0s which we identify based on key range.

Reply:
I think this is fine for now, but the layer can be overly wide.